### ISSUE TITLE
p2p: rate limit number of connection attempts per IP

### ISSUE THREAD

**Issue by melekes** [tendermint/tendermint#4753]
Tendermint does not rate limit P2P connection requests per IP (an attacker can saturate all the inbound slots). See https://github.com/tendermint/tendermint/blob/v0.33/CHANGELOG.md#denial-of-service-1.

**Comment by JayT106** [comment]
@tessr @melekes Is there anyone looking at this issue? If not, I would like to take this. Thank you.

**Comment by melekes** [comment]
> @tessr @melekes Is there anyone looking at this issue? If not, I would like to take this. Thank you.

I don't think so, no. Please do 👍 Feel free to ask any questions and/or post your ideas re planned implementation.

**Comment by JayT106** [comment]
Sure.

**Comment by JayT106** [comment]
@melekes my idea would be to add a simple(and small) LRU-cache in the `Router` to store the `net.IP` of the inbound connection request. Then we define the control rate for the connection request in each IP (would be great if anyone knows what's the preferred rate for the control rate of the connection request), and then check it at https://github.com/tendermint/tendermint/blob/19a375aaabf18c700440eeb4b4c50617fab393a3/p2p/router.go#L455
To use `PeerManager.Accepted()` is good for limiting the connection by NodeID because the PeerManager doesn't know anything about the IP and looks like the NodeID info only knows after a handshake https://github.com/tendermint/tendermint/blob/19a375aaabf18c700440eeb4b4c50617fab393a3/p2p/router.go#L502

**Comment by tomtau** [comment]
is this solved by https://github.com/tendermint/tendermint/pull/6286 ?

**Comment by JayT106** [comment]
> is this solved by #6286?
I think so. I just realized when I try to merge my code today.

**Comment by tychoish** [comment]
Yep! this is now implemented in the new P2P stack. I'm going to go ahead and close this issue.

**Comment by cmwaters** [comment]
Bet you too it :)
